urld58fd5e7a9_longrm_weibo_robot_tgz-pJ8-weibo4j_org_json_TestJ8/src/weibo4j/http/OAuth.java

```diff
@@ -0,0 +1,359 @@
+     * @return encoded string
+     * @see <a href="http://wiki.oauth.net/TestCases">OAuth / TestCases</a>
+     * @see <a href="http://groups.google.com/group/oauth/browse_thread/thread/a8398d0521f4ae3d/9d79b698ab217df2?hl=en&lnk=gst&q=space+encoding#9d79b698ab217df2">Space encoding - OAuth | Google Groups</a>
+     * @see <a href="http://tools.ietf.org/html/rfc3986#section-2.1">RFC 3986 - Uniform Resource Identifier (URI): Generic Syntax - 2.1. Percent-Encoding</a>
+     */
+    public static String encode(String value) {
+        String encoded = null;
+        try {
+            encoded = URLEncoder.encode(value, "UTF-8");
+        } catch (UnsupportedEncodingException ignore) {
+        }
+        StringBuffer buf = new StringBuffer(encoded.length());
+        char focus;
+        for (int i = 0; i < encoded.length(); i++) {
+            focus = encoded.charAt(i);
+            if (focus == '*') {
+                buf.append("%2A");
+            } else if (focus == '+') {
+                buf.append("%20");
+            } else if (focus == '%' && (i + 1) < encoded.length()
+                    && encoded.charAt(i + 1) == '7' && encoded.charAt(i + 2) == 'E') {
+                buf.append('~');
+                i += 2;
+            } else {
+                buf.append(focus);
+            }
+        }
+        return buf.toString();
+    }
+
+    /**
+     * The Signature Base String includes the request absolute URL, tying the signature to a specific endpoint. The URL used in the Signature Base String MUST include the scheme, authority, and path, and MUST exclude the query and fragment as defined by [RFC3986] section 3.<br>
+     * If the absolute request URL is not available to the Service Provider (it is always available to the Consumer), it can be constructed by combining the scheme being used, the HTTP Host header, and the relative HTTP request URL. If the Host header is not available, the Service Provider SHOULD use the host name communicated to the Consumer in the documentation or other means.<br>
+     * The Service Provider SHOULD document the form of URL used in the Signature Base String to avoid ambiguity due to URL normalization. Unless specified, URL scheme and authority MUST be lowercase and include the port number; http default port 80 and https default port 443 MUST be excluded.<br>
+     * <br>
+     * For example, the request:<br>
+     * HTTP://Example.com:80/resource?id=123<br>
+     * Is included in the Signature Base String as:<br>
+     * http://example.com/resource
+     *
+     * @param url the url to be normalized
+     * @return the Signature Base String
+     * @see <a href="http://oauth.net/core/1.0#rfc.section.9.1.2">OAuth Core - 9.1.2.  Construct Request URL</a>
+     */
+
+    public static String constructRequestURL(String url) {
+        int index = url.indexOf("?");
+        if (-1 != index) {
+            url = url.substring(0, index);
+        }
+        int slashIndex = url.indexOf("/", 8);
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

If the indexOf() search on line 292 does not find the substring and returns -1, the call to substring() on line 305 will throw IndexOutOfBoundsException. Note that the method is publicly accessible to external callers who might not be aware of the constraint on string url.